### ISSUE TITLE
Issue trackers organisation

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5#376]
This issue follows somehow https://github.com/ckeditor/ckeditor5/issues/375. It assumes that we'll choose to keep the multi-repo architecture.

I guess that the problems are quite obvious:

* Issues are split into multiple repositories.
* Developers don't know where to report their issues.
* Developers are not able to easily search for issues (e.g. to check whether the problem they have is already reported).
* We'll need to move a lot of issues between repositories.
* Often, a single issue concerns more repositories and then we need to have an "umbrella" ticket somewhere anyway.
* You can either totally disable an issue tracker in a GH repo or keep it open to everyone (which limits the options that we have).
* https://github.com/ckeditor/ckeditor5-design is currently used only for announcements because it was confusing even for us when to use https://github.com/ckeditor/ckeditor5 and when ckeditor5-design.

So far, to manage all these issue trackers, we've been using https://codetree.com which is an amazing tool but, unfortunately, it's not free so our community can't use it. Even if it was free, though, it'd be hard to expect that everyone uses it.

We've discussed these problems briefly and there were three options:

* Keep the status quo – multiple open issue trackers. This is still a possible solution, but we'd need to work on:
  * Documentation (issue templates, explaining where to report if you don't know, how to search for an issue).
  * Cleaning up all the misreported issues. (Note: if you need to move a ticket and you mention the user which reported the original one in the new one, then the user gets notifications from the new one.)
  * Some web app listing all the tickets from all repositories by milestones and perhaps implementing some search mechanisms. Without this it's hard to understand what issues are being covered by which milestone.
  * Pray that all this works in practice.
* Close all issue trackers except the main one. Everything will be conducted from it.
  * It's by far the simplest solution, but it has its downsides.
  * We'll lose all the tickets that we already created (which is ~1000). That's something I'd not accept at the current stage when we had hundreds of important discussions in them. We'd need to move them automatically to ckeditor5.
  * I think that we won't be able to close tickets by `Fixes #XXX.` in PRs.
  * We'd need yet another set of labels (to mark which packages are related to this issue). Minor issue, but still. However, this can also affect automatic changelogs generation (each package has its own one).
  * And last but not least, packages should often be discussed separately. E.g., AFAIK you're not able to watch just a subset of tickets. So you either watch all CKEditor 5 tickets or none. Currently (with multiple issue trackers) you can follow just some of them (which even the core devs do). Also, having all discussions in one place may work towards coupling packages.
* Have one main repo with all the tickets and synchronised discussions in ckeditor5 and package specific trackers for other repositories. This idea requires more explanation and isn't the simplest one but deserves to be discussed. It requires developing a bot synchronising issues between the main repository and package repos.
  * The idea is to have everyone (except perhaps the core team) reporting tickets in ckeditor5. This will ensure that everyone will know where to report tickets and that all tickets can be found there. This is especially important for some bug reports where a stack trace or error name is listed.
  * Once ticket is verified, a related ticket will be opened by the core team in the specific repo(s). By using some special references (e.g. in the package ticket one will write "Parent: ckeditor#NNN"), the bot will logically join these tickets and once the package ticket is closed by a PR in this repo, the bot will close the parent ticket.
  * I'm not sure what should happen if a core dev opens a ticket in a package repo. Perhaps some of them should be synced up to ckeditor5 (ones which make sense there). The bot should also write a comment under a parent ticket (if there's any) so the developers watching that original one will be notified that the core team began work on this issue.
  * This will:
    * allow us to have meaningful and package-scoped discussions in specific packages while maintaining one "entry point" for the community,
    * allow us to work on some package specific maintenance (like refactoring) without polluting the main repo,
    * make it possible to have multiple sub-tickets connected to one parent ticket in ckeditor5 (the parent ticket would only be closed if all sub-tickets are closed),
    * have a clean ticket in the package repo which will make for a good starting point for a changelog (issues reported by the community has usually quite bad titles),
    * allow closing multiple parent tickets using one ticket (e.g. multiple bugs by one patch).

I certainly like the last option the most. It feels the most powerful and satisfies all requirements. It also assumes adding what GitHub misses – relations between tickets. However, it will require writing a bot so it requires far more work than e.g. the second option.

## Off topic – what about ckeditor5-design

This is a separate topic, but it's time to conclude this one too. It doesn't make sense to keep this repo as a "design discussions". It's both confusing (is it about UI/UX?) and superfluous (we've got ckeditor5 repo for general discussions, like this one :D).

I see three options:

* Rename it to `ckeditor5-news` (because it's only used in this way now).
* Close it and open a Disqus forum instead.
* Introduce some other news platform to avoid confusion what is it for.

**Comment by fredck** [comment]
I'm for "Keep the status quo" without doing any other move other than including documentation for where to report issues. Later on, we'll have a better feeling if any further work is necessary to enhance this.

**Comment by Reinmar** [comment]
Hm... It seems that if we keep using issues the way we did in the past, then the last solution that I proposed will be a possible extension of what we have. So that can be the best way to save now some time and not regret something in the future.

**Comment by wwalc** [comment]
Some idea to solve the headache with moving issues: maybe a browser extension could enhance the GitHub Experience for us (in the future)? 

The extension could add text input field with autocomplete below "Labels", where by typing a part of the name of ckeditor5 repository it would offer the list of choices (or let it even be a select element at the beginning). After choosing the target repository, the extension would automate the following:
 - creating a new ticket in the correct repository (on behalf of the user with installed extension),
 - mentioning the original reporter in the copy of an issue so that he got notified on future updates
 - copying all comments in a similar manner (mentioning authors).

**Comment by wwalc** [comment]
Lol I just googled out that most likely we do not even have to create this: https://github.com/google/github-issue-mover

**Comment by Reinmar** [comment]
I wouldn't rely on users installing any extensions. 1% of them will. I wouldn't :D Either we'll automate this or we can't expect anything.

**Comment by wwalc** [comment]
I was thinking mainly about reducing the amount of work **for us** with moving issues reported in random places.

**Comment by Reinmar** [comment]
Ah, ok. So a tool for that would be nice. But OTOH, if 99% of users will report tickets in the main repo (as we'll recommend in the issue template), then moving the other 1% manually won't be a big deal. I would not move issues to package repos anyway – in such case we can (and should) only mention in the original report that we started working on a solution for that issue. We'd close both issues once PR is merged. Just like in the 3rd option, but without automation.

**Comment by wwalc** [comment]
I agree, the issue template pointing users to the main repo should already help much. No need to hurry with the extension, just wanted to note it down somewhere in case we were flooded with tickets landing in a wrong place.

**Comment by Reinmar** [comment]
BTW, and what about ckeditor5-design?

/cc @wojtekidd @AnnaTomanek

**Comment by wojtekidd** [comment]
ckeditor5-news or ckeditor5-updates would be fine with me.

**Comment by fredck** [comment]
> what about ckeditor5-design

> It's both confusing (is it about UI/UX?)

Never heard anyone other than us saying this.

> superfluous (we've got ckeditor5 repo for general discussions, like this one :D).

Now maybe yes. I think that the design repo fulfilled its goal, to bootstrap the design of the editor. It was about discussing things that still didn't have a place where to discuss them. For instance, right now, it makes sense talking about new design topics around the engine at https://github.com/ckeditor/ckeditor5-engine.

As we're reaching a stable design, we'll have less of such talks, anyway.

Other than that, little people would be interested in participating in such discussions. Generally, those who do it will be commenting in already open topics. This is mainly for the core team and so we can organize it in the way we want.

The really important thing is our ability of moving the final results of our design decisions and implementation into documentation. That's the place where people will really like to see such things.

**Comment by Reinmar** [comment]
> Never heard anyone other than us saying this.

Fact that we haven't heard that doesn't mean that isn't confusing. Don't expect issue reports about UX matters. But of course, even if it is confusing, this isn't super important.

> I think that the design repo fulfilled its goal, to bootstrap the design of the editor

IMO, we made a mistake and should've kept those things in one place from the day one. And since it doesn't make much sense any more, I raised the question what to do with it now. We must start thinking about cleaning the mess that we did during development and this is one of these things which should be reviewed. Especially that we've also see a need for a different communication channel.

**Comment by fredck** [comment]
We could still use design for discussion like the one we're having right now. But for some unclear reason, in the middle of the road, that repo has been abandoned. It is clear that there is an intention to close it, so let's move forward with it.

Design could remain there, for history purposes. We would still be able to point to its issues when trying to remember or to justify our decisions. Finally, a good part of it would be transferred to documentation.

**Comment by Reinmar** [comment]
> But for some unclear reason, in the middle of the road, that repo has been abandoned

It was abandoned because there's no clear distinction between what's design and what's not. Having the design process in one place and implementation in another turned to be unrealistic, because it assumes that after the design phase there's implementation and nothing more. That's rarely true in an iterative way of working.

E.g. we could see this perfectly with the wiki pages that we created. They would have to be constantly updated because nearly nothing from the original design is now true... Except the code style :D.

Of course, I'm for keeping the repo because the issues should stay there for reference. However, that's not what I was asking, because, as you noticed yourself, the decision has been made a long time ago. What's still open is where to go with news and planning.

But I just noticed one thing – since this repo must stay open anyway and we lack the "planning space", maybe we could just renamed it to `ckeditor5-planning`. Then even the initial topics would make some sense. WDYT?

**Comment by fredck** [comment]
Well, then `ckeditor5` would fit well anyway as it centralizes everything.

**Comment by Reinmar** [comment]
> Well, then ckeditor5 would fit well anyway as it centralizes everything.

Except that you're not able to follow just a single label on GH :(

**Comment by oleq** [comment]
I stumbled across this issue and it looks like it has been resolved. `ckeditor5` repo centralizes our project and works as an entry point for submitting issues. 

I think it's time we closed it. WDYT?
